onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -34,6 +34,55 @@ REGISTER_KERNEL_TYPED(float)
 REGISTER_KERNEL_TYPED(double)
 REGISTER_KERNEL_TYPED(MLFloat16)
 
+template <typename T>
+const cudnnConvolutionFwdAlgo_t Conv<T>::kAllAlgos[] = {
+    CUDNN_CONVOLUTION_FWD_ALGO_GEMM,
+    CUDNN_CONVOLUTION_FWD_ALGO_FFT,
+    CUDNN_CONVOLUTION_FWD_ALGO_FFT_TILING,
+    CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_GEMM,
+    CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM,
+    CUDNN_CONVOLUTION_FWD_ALGO_DIRECT,
+    CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD,
+    CUDNN_CONVOLUTION_FWD_ALGO_WINOGRAD_NONFUSED,
+};
+
+cudnnStatus_t getWorkspaceSize(const CudnnConvState<cudnnConvolutionFwdAlgoPerf_t>& s,
+                               cudnnConvolutionFwdAlgo_t algo, size_t* sz) {
+  return cudnnGetConvolutionForwardWorkspaceSize(
+      s.handle,
+      s.x_tensor,
+      s.w_desc,
+      s.conv_desc,
+      s.y_tensor,
+      algo,
+      sz);
+}
+
+template <typename algo_t>
+size_t getMaxWorkspaceSize(const CudnnConvState<cudnnConvolutionFwdAlgoPerf_t>& s,
+                           const algo_t* algo, int n_algo) {
+  size_t max_ws_size = 0;
+
+  // TODO: get maximum available size from memory areana
+
+  size_t free, total;
+  CUDA_CALL_THROW(cudaMemGetInfo(&free, &total));
+  // Assuming 10% of fragmentation
+  free = static_cast<size_t>(static_cast<double>(free) * 0.9);
+
+  std::cout << "free: " << free << " total: " << total << std::endl;
+
+  for (int i = 0; i < n_algo; i++) {
```

Comment:
Is this cudnnGetConvolutionForwardWorkspaceSize an expensive function? If so, would it cost too much to run it for each conv kernel? Is there other speculative way to avoid this check, e.g., set a relatively large default workspace size? Or if it throws  an exception when workspace is too small, could we catch it and enlarge the workspace size?

Reply:
How does PyTorch does it? Do they cache workspace size results?

Reply:
@duli2012, cudnnGetConvolutionForwardWorkspaceSize should be host-only CPU call. If we can't find the cached algorithm for current input shape, we will go through the algo search process. 

@vadimkantorov, PyTorch does the same process for cache miss cases.